src/ape_cache/query.py

```diff
@@ -0,0 +1,472 @@
+                check of the number of rows that match the clause.
+
+        Raises:
+            :class:`~ape.exceptions.QueryEngineError`: When given an
+                incompatible QueryType
+
+        Returns:
+            `sqlalchemy.sql.expression.TextClause`
+        """
+
+        raise QueryEngineError(
+            "Not a compatible QueryType. For more details see our docs "
+            "https://docs.apeworx.io/ape/stable/methoddocs/"
+            "exceptions.html#ape.exceptions.QueryEngineError"
+        )
+
+    @_perform_query_clause.register
+    def _perform_block_clause(self, query: BlockQuery) -> Select:
+        return (
+            select([column(c) for c in query.columns])
+            .where(Blocks.number >= query.start_block)
+            .where(Blocks.number <= query.stop_block)
+            .where(Blocks.number % query.step == 0)
+        )
+
+    @_perform_query_clause.register
+    def _perform_transaction_clause(self, query: BlockTransactionQuery) -> Select:
+        return select([Transactions]).where(Transactions.block_hash == query.block_id)
+
+    @_perform_query_clause.register
+    def _perform_contract_event_clause(self, query: ContractEventQuery) -> Select:
+        return (
+            select([column(c) for c in query.columns])
+            .where(ContractEvents.block_number >= query.start_block)
+            .where(ContractEvents.block_number <= query.stop_block)
+            .where(ContractEvents.block_number % query.step == 0)
+        )
+
+    @singledispatchmethod
+    def perform_query(self, query: QueryType) -> Optional[Iterator]:  # type: ignore
+        """
+        Performs the requested query from cache.
+
+        Args:
+            query (QueryType): Choice of query type to perform a
+                check of the number of rows that match the clause.
+
+        Raises:
+            :class:`~ape.exceptions.QueryEngineError`: When given an
+                incompatible QueryType
+
+        Returns:
+            Optional[Iterator]
+        """
+
+        raise QueryEngineError(
+            "Not a compatible QueryType. For more details see our docs "
+            "https://docs.apeworx.io/ape/stable/methoddocs/"
+            "exceptions.html#ape.exceptions.QueryEngineError"
+        )
+
+    @perform_query.register
+    def _perform_block_query(self, query: BlockQuery) -> Optional[Iterator[BlockAPI]]:
+        try:
+            with self.database_connection as conn:
+                result = conn.execute(self._perform_query_clause(query))
+
+                if not result:
+                    # NOTE: Should be unreachable if estimated correctly
+                    raise QueryEngineError(f"Could not perform query:\n{query}")
+
+                for row in result:
+                    yield self.provider.network.ecosystem.decode_block(dict(row.items()))
+
+        except QueryEngineError as err:
+            logger.error(f"Database not initiated: {str(err)}")
```

Comment:
I think this should actually just raise any issues here
```suggestion
```

Also note that it's impossible to make it this fair un-initialized (without some sort of corruption)

Reply:
[21465f7](https://github.com/ApeWorX/ape/pull/680/commits/21465f70f97a56ad08ed00e35500fc40864eef1d)

This is no longer required, we removed the try-except